[rush]  Use environment variable to pass publishing token

darsi-an opened the pull request:
This is a WIP to close #1588. 

I've updated this PR following the design suggested below. Please review!

octogonz commented:
Thanks very much for working on this!

octogonz commented on apps/rush-lib/src/cli/actions/PublishAction.ts:
```diff
@@ -487,7 +488,7 @@ export class PublishAction extends BaseRushAction {
 
     if (this._npmAuthToken.value) {
-      args.push(`--${registry}:_authToken=${this._npmAuthToken.value}`);
+      args.push(`--${registry}:_authToken=\$\{${this._npmAuthToken.environmentVariable}\}`);
```
This is still going to show up in the build log I think.  

@iclanton  Do you know if there is a way to pass tokens to PNPM/NPM via an environment variable, instead of on the command-line or npmrc file?

darsi-an replied:
I was unsure about this too. My VS CODE output while debugging publish was: `--//registry.npmjs.org/:_authToken=${NPM_TOKEN} --access public (apps\rush-publish-test)`. I'm not using a local .npmrc file, and just have NPM_TOKEN set in my machine's environment variables. I have tested this multiple times, but, relying on the output via VS code. I'm not sure on how other systems would handle it.

darsi-an replied:
Yeah you're right. The current code only allowed me to publish using the environmentVariable property, it shouldnt be visible. Id love to hear more about this if anyone has some information, ill update this as soon as possible.

octogonz replied:
[These instructions](https://blog.npmjs.org/post/118393368555/deploying-with-npm-private-modules) and [this thread](https://npm.community/t/cannot-set-npm-config-keys-containing-underscores-registry-auth-tokens-for-example-via-npm-config-environment-variables/233/13) suggest that you have to invent your own environment variable name, and put it in an `.npmrc` file, like this:

**.npmrc**
```
//registry.npmjs.org/:_authToken=${MY_AUTH_TOKEN}
```

Then the CI machine can specify `MY_AUTH_TOKEN` as an environment variable when running `npm publish`.  

However, there is a problem that we probably DON'T want to specify this variable during normal working.  For example, on GitHub for normal development we want to have:

**common/config/rush/.npmrc**
```
registry=https://registry.npmjs.org/
always-auth=false
```

...whereas in a private repo with a private registry, we'd want something like:

**common/config/rush/.npmrc**
```
always-auth=true
```
**~/.npmrc**
```

octogonz replied:
I chatted with @iclanton  about this.  It seems that the line must go in an `.npmrc` file. According to the [npmrc docs](https://docs.npmjs.com/files/npmrc), there are only 4 possible locations for this file.  In order of precedence:

- per-project config file (/path/to/my/project/.npmrc)
- per-user config file (~/.npmrc)
- global config file ($PREFIX/etc/npmrc)
- npm builtin config file (/path/to/npm/npmrc)

We considered writing `/path/to/my/project/.npmrc` before `npm publish` and deleting it afterwards, but that can leave an unclean state if the CI job gets cancelled.  Also, there may be valid cases for user-defined settings in this file location.

So instead we are thinking Rush should use `~/.npmrc`.  The home folder can get redirected fairly easily.  For example, on Windows if you do:

```
$ set USERPROFILE=C:\Git\rushstack\common\temp\publish-home
$ cd C:\Git\rushstack\my\project
$ npm config list
```
...then it will print out settings found in `C:\Git\rushstack\common\temp\publish-home\.npmrc`.

A possible risk is that `npm publish` might somehow depend on other state from the home directory, but this seems unlikely.  Note that we would only be redirecting `USERPROFILE` during that specific command `npm publish`.  The `rush install` and `rush build` stuff would have access to the normal home directory.

Where does `common\temp\publish-home\.npmrc` come from?  We're proposing that the user would create a file `common\config\rush\.npmrc-publish`, and add it to Git.  Rush would copy it to `common\temp\publish-home` prior to `rush publish`.

For backwards compatibility, if `.npmrc-publish` does not exist, then we would fall back to the old behavior.

@darsi-an How does that sound?  It seems like probably not too much work to implement.  The only slightly tricky part is we would probably need separate logic for redirecting the home directory on each OS (Mac/Linux/Windows).  

If you're able to implement+test the file copying idea, I can help with the documentation and `rush init` part.

darsi-an replied:
Sounds good, I like the approach. I'll update my PR accordingly. Regarding the `rush init` part, is it to generate a `common\config\rush\.npmrc-publish` file?

darsi-an commented:
@octogonz, so I've updated my PR to support the design suggested by you and @iclanton. I am able to copy `common\config\rush\.npmrc-publish` to `\common\temp\publish-home` and publish to npm registry using its auth token. Additionally, I am able to redirect the npm userconfig setting by updating the env variable for $HOME or $USERPROFILE, prior to passing the environment to the execute command. I've also added an optional parameter to Utilities syncNpmrc to support the .npmrc-publish file. 

I have tested this on Windows and Linux (ubuntu) and it works as expected. In regards to the actual tests, do I need to add test cases for syncNpmrcPublish in PublishUtilities.test? The function does not return anything, so I would be just checking if the `common\temp\publish-home\.npmrc" file exists. Please correct me if I am wrong.

Currently if any environment variables specified in a .npmrc file are not set, it will remove the line and add a comment to `common\temp\.npmrc`, such as `"; MISSING ENVIRONMENT VARIABLE: //my-registry.com/npm/:_authToken=${MY_AUTH_TOKEN}"`. I came across a [comment](https://github.com/microsoft/rushstack/issues/1452#issuecomment-519680211 ) of yours and was wondering if we should in fact throw an error, as no valid authentication token will return with exit code 1. However a user can check the .npmrc in `common\temp` or `common\temp\publish-home` to remedy this.

Trying to publish with an unset environment variable will fail, unless you pass a valid `--npm-auth-token` to the command. Is it safe to assume that setting  `--npm-auth-token` via command line, takes precedence over any .npmrc file? From testing this situation it seems as it does.

If there is anything more that is required, please let me know!

octogonz commented:
Awesome, thanks very much! We'll take a look ASAP

octogonz commented:
> In regards to the actual tests, do I need to add test cases for syncNpmrcPublish in PublishUtilities.test? The function does not return anything, so I would be just checking if the `common\temp\publish-home\.npmrc` file exists. Please correct me if I am wrong.

I'm not too worried about it. The success of publishing operations mainly involves actual files on disk and the behavior of the `npm` command, so manual testing is probably more useful than unit tests.

> Currently if any environment variables specified in a .npmrc file are not set, it will remove the line and add a comment to `common\temp\.npmrc`, such as `"; MISSING ENVIRONMENT VARIABLE: //my-registry.com/npm/:_authToken=${MY_AUTH_TOKEN}"`. 

If I remember right, we do this because the NPM tool will otherwise (stupidly) leave the `${MY_AUTH_TOKEN}` string unexpanded, which causes confusing errors.  Whereas if the line is commented out, it will work correctly in situations where an explicit token is not needed. 

> I came across a [comment](https://github.com/microsoft/rushstack/issues/1452#issuecomment-519680211) of yours and was wondering if we should in fact throw an error, as no valid authentication token will return with exit code 1.

This is a good idea, but maybe we should consider it as a separate PR.  I feel like the validation rules will be slightly tricky (e.g. there can be multiple environment variables), and it's questionable how much it will really help people. Setting up publishing is done once, usually by a somewhat knowledgeable person, and this is just one of many things that can go wrong.

> Is it safe to assume that setting `--npm-auth-token` via command line, takes precedence over any .npmrc file? From testing this situation it seems as it does.

This makes sense. In the future I'd like to deprecate `--npm-auth-token` and eventually remove it entirely.

octogonz commented:
This PR looks good.  The optional parameter for `syncNpmrc()` seemed a little awkward, so I pushed https://github.com/microsoft/rushstack/pull/1602/commits/edfe7aff025a6dd0db70486d58b3fec41d90dfb3 to your branch. It refactors the environment variable expansion into a separate function  `copyAndTrimNpmrcFile()` which accepts the full path `.npmrc-publish`.  I think it's also a bit cleaner, since `syncNpmrc()`'s "syncing" (i.e. deleting the target if the source is missing) didn't seem very useful when the source/target folders have different `.npmrc` filenames.

octogonz commented:
Regarding this `TODO`...

```ts
  public static execCommand(
    shouldExecute: boolean,
    command: string,
    args: string[] = [],
    workingDirectory: string = process.cwd(),
    environment?: IEnvironment
  ): void {

    let relativeDirectory: string = path.relative(process.cwd(), workingDirectory);

    if (relativeDirectory) {
      relativeDirectory = `(${relativeDirectory})`;
    }

    let commandArgs: string = args.join(' ');

    if (args.includes('publish')) {
      // Remove token from printing to output
      // TODO: Find the best approach   <=====================
      const regex: RegExp = /=.*?\s/;
      commandArgs = commandArgs.replace(regex, ' ');
    }
```
...from looking at the code, it seems that the only two operations that rely on `--npm-auth-token` are [PublishAction._npmPublish()](https://github.com/microsoft/rushstack/blob/edfe7aff025a6dd0db70486d58b3fec41d90dfb3/apps/rush-lib/src/cli/actions/PublishAction.ts#L370) and [PublishAction._packageExists()](https://github.com/microsoft/rushstack/blob/edfe7aff025a6dd0db70486d58b3fec41d90dfb3/apps/rush-lib/src/cli/actions/PublishAction.ts#L431).  Seems like we could simply implement custom logging for those functions rather than relying on `PublishUtilities.execCommand()`'s logging.

Hmm.... I'm also thinking that your `.npmrc-publish` file needs to get copied *before* the first call to `PublishAction._packageExists()`.

octogonz commented:
> Hmm.... I'm also thinking that your `.npmrc-publish` file needs to get copied _before_ the first call to `PublishAction._packageExists()`.

This is an important piece that we need to solve before we can merge this PR.  Because `PublishAction._packageExists()` calls `_addSharedNpmConfig()`, it is relying on the `--npm-auth-token` value.  Thue `_packageExists()` definitely needs to support  your `.npmrc-publish` feature as well.  And we probably should write that file only once at the start of the operation, rather than repeatedly rewriting the file for each call.

It should be a minor fix, but I'm out of time for today.

darsi-an commented:
> > Hmm.... I'm also thinking that your `.npmrc-publish` file needs to get copied _before_ the first call to `PublishAction._packageExists()`.
> 
> This is an important piece that we need to solve before we can merge this PR. Because `PublishAction._packageExists()` calls `_addSharedNpmConfig()`, it is relying on the `--npm-auth-token` value. Thue `_packageExists()` definitely needs to support your `.npmrc-publish` feature as well. And we probably should write that file only once at the start of the operation, rather than repeatedly rewriting the file for each call.
> 

Yeah you're right, `.npmrc-publish` should get copied before the first call to `PublishAction._packageExists()`. The changes I've pushed are a bit different from the quote above.

Instead, I added a function [_addNpmPublishHome()](https://github.com/microsoft/rushstack/blob/959bc5579e7c0629a992bc832d964294825566c5/apps/rush-lib/src/cli/actions/PublishAction.ts#L486) which creates `"common\temp\publish-home\.npmrc"` if `common\config\rush\.npmrc-publish` exists. This function is called after validating publish command-line parameters, before, any calls to `PublishAction._packageExists()`. 

Additionally, I've updated `_addSharedNpmConfig` to simply check if the `"common\temp\publish-home\.npmrc"` exists and apply the appropriate redirects. Rather than extending the feature to `PublishAction._packageExists()`, I think this is more appropriate because `_addSharedNpmConfig` already [redirects ](https://github.com/microsoft/rushstack/blob/959bc5579e7c0629a992bc832d964294825566c5/apps/rush-lib/src/cli/actions/PublishAction.ts#L525-L532)(and pushes to args) config tokens specified via command-line parameters. 

If there is a reason to extend the functionality to `PublishAction._packageExists()`, or any other changes that need to be made please let me know, I will update the code.

octogonz commented:
@darsi-an This seems reasonable. One thought: is the behavior always deterministic? E.g. are there cases where the package manager can accidentally use a stray common/temp/.npmrc file left behind from a previous Rush command? If so maybe we should always delete the file if we don't copy it. Currently it is (fairly reliably) whatever would be left behind from `rush install`.

octogonz commented:
BTW I know of a repo that does `rush install ` against an internal registry before building, but then does `rush publish` against a different registry for the release. Thus it's important that we use the right credentials and registry URL when checking whether the package is already published.

darsi-an commented:
> One thought: is the behavior always deterministic?

Good point. If `.npmrc-publish` does not exists it should fall back to old behavior (publish based on environmental npm registry values or the command-line parameters). There should be no redirect of `'userconfig'` to `common/temp/.npmrc` during rush publish. Doing so would eliminate the possibility of accidentally using a stray `common/temp/.npmrc` file when publishing. However, we would need to delete stray `common/temp/publish-home/.npmrc` file if `.npmrc-publish` doesn’t exist.

darsi-an commented:
It seems that stray `common/temp/.npmrc` file would not override config settings from `userconfig`. Per-project config file would only be resolved at the highest priority if npm were running at the root, in our case in `common/temp`. From what I can tell after debugging and testing different cases, `rush publish` executes a package manager’s `publish` action passing the path of the package being published as the working directory. In contrast, `rush install` copies over the `common/config/rush/.npmrc` and executes a package manager’s `install` action passing `common/temp` as the working directory.

darsi-an commented:
### Few points for discussion:

**W.R.T #1665**

> If `.npmrc` has `registry=https://mycustomregistry.example.org` NPM will publish packages to that registry and not the default (https://registry.npmjs.org).

I think he's referring to the per-project config file. This is the correct behavior of `npm`, if `/path/to/my/project/.npmrc` is present. 

**In my last comment I noted:**
> Per-project config file would only be resolved at the highest priority if npm were running at the root, in our case in `common/temp`

What led me to this was, _How would a package manager use a stray `common/temp/.npmrc` file left behind?_

**From npmrc** [**docs**](https://docs.npmjs.com/files/npmrc.html#per-project-config-file)
 > Per-project config file
When working locally in a project, a .npmrc file in the root of the project (ie, a sibling of node_modules and package.json) will set config values specific to this project. Note that this only applies to the root of the project that you’re running npm in. 

> So the fix would need to support that.

This may already be 'supported'. I was able to publish a package using per-project config file by just adding .npmrc (with configs) at the projects root. I believe this is b/c `rush publish` runs `package-manager publish` in the appropriate directories, so `.npmrc` at project's root would be resolved at the highest priority.

However If we were to support this, users could have multiple `npmrc`  files (one for every project)? Might be hard to manage and track all those files. 

**Other considerations/cases I still need to verify** 
- Effect of .npmrc used when bootstrapping rush for CI jobs
- Do we handle package.json [publishconfig](https://docs.npmjs.com/files/package.json#publishconfig)

iclanton commented on apps/rush-lib/src/scripts/install-run.ts:
```diff
@@ -63,45 +62,57 @@ function _parsePackageSpecifier(rawPackageSpecifier: string): IPackageSpecifier
  * home directory.
  *
- * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH Utilities._syncNpmrc()
+ * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH _copyNpmrcFile() FROM scripts/install-run.ts
```
```suggestion
 * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH Utilities._copyNpmrcFile()
```

iclanton commented on apps/rush-lib/src/scripts/install-run.ts:
```diff
@@ -63,45 +62,57 @@ function _parsePackageSpecifier(rawPackageSpecifier: string): IPackageSpecifier
+      _copyAndTrimNpmrcFile(sourceNpmrcPath, targetNpmrcPath);
     } else if (fs.existsSync(targetNpmrcPath)) {
       // If the source .npmrc doesn't exist and there is one in the target, delete the one in the target
+      console.log(`Deleting ${targetNpmrcPath}`);
```
Is this log message useful?

darsi-an replied:
The earlier  `_syncNpmrc()` in install-run script did not have log messages. Is there a specific reason why (maybe for CI purposes)?

iclanton replied:
I guess they aren't doing any harm. Can you add a `// Verbose` comment to these, so when we replace the logger in Rush with `Terminal`, we'll know to redirect these to verbose?

```suggestion
      console.log(`Deleting ${targetNpmrcPath}`); // Verbose
```

darsi-an replied:
Done.

iclanton commented on apps/rush-lib/src/scripts/install-run.ts:
```diff
@@ -63,45 +62,57 @@ function _parsePackageSpecifier(rawPackageSpecifier: string): IPackageSpecifier
+ * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH _copyNpmrcFile() FROM scripts/install-run.ts
+ */
+function _copyAndTrimNpmrcFile(sourceNpmrcPath: string, targetNpmrcPath: string): void {
+  console.log(`Copying ${sourceNpmrcPath} --> ${targetNpmrcPath}`);
```
Is this log message useful?

iclanton commented on apps/rush-lib/src/cli/actions/PublishAction.ts:
```diff
@@ -478,8 +483,51 @@ export class PublishAction extends BaseRushAction {
+      }
+    } catch (e) {
+      throw new Error(`Error syncing .npmrc-publish file: ${e}`);
+    }
```
This code seems very similar to `Utilities.syncNpmrc`. Can we just add a parameter to that function for the `.npmrc` file name and deduplicate this logic (and maybe revert the refactor to that code to reduce churn)?

darsi-an replied:
> (and maybe revert the refactor to that code to reduce churn)?

Are you referring to the optional Boolean flag in `syncNpmrc()` I had in [`261e44f`](https://github.com/microsoft/rushstack/pull/1602/commits/261e44f74ce5992e2ab405e9c3f9f6b52238a03f)?

iclanton replied:
I was referring to the change that extracted the a bunch of the functionality into a function.

darsi-an replied:
Just to be on the safe side, my previous commit, [`b25306c`](https://github.com/microsoft/rushstack/pull/1602/commits/b25306cdb4d6e638c6700799b02d162328095d70), that duplicated logic in `syncNpmrc()` logic?

iclanton replied:
I was actually referring to this commit https://github.com/microsoft/rushstack/commit/edfe7aff025a6dd0db70486d58b3fec41d90dfb3, but in retrospect I'm cool with the refactor. Let's just deduplicate the sync code.

darsi-an replied:
Okay, sounds good.

iclanton commented on apps/rush-lib/src/cli/actions/PublishAction.ts:
```diff
@@ -478,8 +483,51 @@ export class PublishAction extends BaseRushAction {
+    const targetNpmrcPublishFolder: string = path.join(this.rushConfiguration.commonTempFolder, 'publish-home');
+
+    // Example: "common\temp\publish-home\.npmrc"
+    const targetNpmrcPublishPath: string = path.join(targetNpmrcPublishFolder, '.npmrc');
```
These were defined in the above function. These should probably just be private instance fields.

iclanton commented on apps/rush-lib/src/cli/actions/PublishAction.ts:
```diff
@@ -478,8 +483,51 @@ export class PublishAction extends BaseRushAction {
+  }
+
   private _addSharedNpmConfig(env: { [key: string]: string | undefined }, args: string[]): void {
+    const userConfig: string = (process.platform === 'win32') ? 'USERPROFILE' : 'HOME';
```
```suggestion
    const userHomeEnvVariable: string = (process.platform === 'win32') ? 'USERPROFILE' : 'HOME';
```

iclanton commented on common/changes/@microsoft/rush/useEnvVariableToPassPublishToken_2019-10-30-10-10.json:
```diff
@@ -0,0 +1,11 @@
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Improve security by enabling the \"rush publish\" auth token to be specified via an environment variable in common/config/rush/.npmrc-publish",
```
```suggestion
      "comment": "Improve security by allowing the \"rush publish\" authentication token to be specified via an environment variable.",
```

iclanton commented on apps/rush-lib/src/cli/actions/PublishAction.ts:
```diff
@@ -499,7 +501,7 @@ export class PublishAction extends BaseRushAction {
 
     // Example: "common\temp\publish-home\.npmrc"
-    const targetNpmrcPublishPath: string = path.join(targetNpmrcPublishFolder, '.npmrc');
+    this._targetNpmrcPublishPath = path.join(targetNpmrcPublishFolder, '.npmrc');
```
Do this in the constructor.

iclanton commented on apps/rush-lib/src/cli/actions/PublishAction.ts:
```diff
@@ -486,8 +493,46 @@ export class PublishAction extends BaseRushAction {
+    const targetNpmrcPublishFolder: string = path.join(this.rushConfiguration.commonTempFolder, 'publish-home');
+
+    // Example: "common\temp\publish-home\.npmrc"
+    this._targetNpmrcPublishPath = path.join(targetNpmrcPublishFolder, '.npmrc');
```
Can you do this directly in the constructor so it's obvious that it's set?

darsi-an replied:
I was originally going to initialize this in the constructor. But to stay consistent with other rush-lib cli actions I set it in `_addNpmPublishHome` . Currently `ChangeAction`, `InitAction`, `PublishAction`, `VersionAction` are the only actions I see that declare private instance fields that are not `ts-command-line` parameters and all these are not done in the constructor.  I can do it directly in the constructor if you'd like, just noted my reasoning.

iclanton replied:
I think I'd rather it be in the constructor. We should probably go through the other actions and fix those up eventually.

darsi-an replied:
Updated with changes. I made `targetNpmrcPublishFolder` private given that `_addSharedNpmConfig()` gets called multiple times and its use in `_addNpmPublishHome`.

iclanton commented on apps/rush-lib/src/scripts/install-run.ts:
```diff
@@ -63,45 +62,57 @@ function _parsePackageSpecifier(rawPackageSpecifier: string): IPackageSpecifier
+ * syncNpmrc() copies the .npmrc file to the target folder, and also trims unusable lines from the .npmrc file.
+ * If the source .npmrc file not exist, then syncNpmrc() will delete an .npmrc that is found in the target folder.
+ *
+ * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH _syncNpmrc() FROM scripts/install-run.ts
```
```suggestion
 * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH Utilities._syncNpmrc()
```

iclanton commented on apps/rush-lib/src/utilities/Utilities.ts:
```diff
@@ -513,44 +512,54 @@ export class Utilities {
+   * IMPORTANT: THIS CODE SHOULD BE KEPT UP TO DATE WITH _copyNpmrcFile() FROM scripts/install-run.ts
+   */
+  public static copyAndTrimNpmrcFile(sourceNpmrcPath: string, targetNpmrcPath: string): void {
+    console.log(`Copying ${sourceNpmrcPath} --> ${targetNpmrcPath}`);
```
```suggestion
    console.log(`Copying ${sourceNpmrcPath} --> ${targetNpmrcPath}`); // Verbose
```

iclanton commented on apps/rush-lib/src/utilities/Utilities.ts:
```diff
@@ -513,44 +512,54 @@ export class Utilities {
+        Utilities.copyAndTrimNpmrcFile(sourceNpmrcPath, targetNpmrcPath);
       } else if (FileSystem.exists(targetNpmrcPath)) {
         // If the source .npmrc doesn't exist and there is one in the target, delete the one in the target
         console.log(`Deleting ${targetNpmrcPath}`);
```
```suggestion
        console.log(`Deleting ${targetNpmrcPath}`); // Verbose
```

iclanton commented on apps/rush-lib/src/cli/actions/PublishAction.ts:
```diff
@@ -486,8 +500,39 @@ export class PublishAction extends BaseRushAction {
+      }
+    } catch (e) {
+      throw new Error(`Error syncing .npmrc-publish file: ${e}`);
+    }
```
You should be able to replace most of this with `Utilities.syncNpmrc`

octogonz commented:
This was published with Rush 5.20.0 and [is now in use](https://github.com/microsoft/rushstack/blob/master/common/config/rush/.npmrc-publish) by the Rush Stack repo. Thanks again @darsi-an !
